design-proposals/2022-10-27-rendering-modularization.md

```diff
@@ -0,0 +1,508 @@
+
+### Optimization
+
+_Addresses evaluation metrics [#2](#eval), [#4](#eval), [#5](#eval), and [#6](#eval)._
+
+A number of our goals are related to improving the performance of the toolkit or at least not making it much worse.  I'd propose that we try not to make it worse at all and even improve it in a few areas if the opportunity arises.
+
+For this PR let's restrict that to areas where we're likely to have made it worse and need to improve something to compensate.  For example, if we make the toolkit larger, we should try to make it smaller in another way.  If we make it slower, we'll try to make it faster to compensate.
+
+Before taking any action, the developers should generate a report on the three key areas:
+- Did the toolkit get larger on iOS or Android?
+- Did the frame rate go down on either platform?
+- Is the toolkit taking longer to startup on either platform?
+
+If the answer to any of these questions is 'yes', then we move to the next stage.
+
+#### Required Changes
+
+These are contingent on what we may have made worse and why.  Opportunities to improve that specific situation will be strategic and hard to pin down at this point.  A lot depends on why the toolkit is bigger, slower, etc...  So this is speculation right now.
+
+It depends on the other changes, but let's break it down by type.
+- Is the toolkit slower to render?  Look at the frame logic and do more up front or on other threads.
+- Is the toolkit slower to startup?  Figure out why and consider doing more work on separate threads or later.
+- Is the toolkit bigger?  Look at slicing unused layers out with compile flags, getting rid of unused code or dependencies.
+
+#### Benefits
+
+The main benefit is just keeping the performance the same.  When we implement Metal on iOS we can expect rendering performance to get much better, of course, but that's not the goal of this effort.  We aim to not make it any worse while reaching our other goals.
+
+While we're doing that, we may run across areas for improvement.  I'd suggest that the developers write these up as they find them.  Some obvious improvements are:
+- Slicing out unused functionality
+- Making more specific functionality optional
+- Reducing or eliminating dependencies on third party libraries.
+
+There's a lot that could be done in general and I'd encourage sponsors to pay maintainers to look into those.
+
+## API Modifications
+
+At present we do not expect to change any existing APIs at the public level.  We do expect to add some new API methods and classes.  We would detail these as we proceeded with the implementation.
+
+## Migration Plan and Compatibility
+
+We do not intend for these changes to be incompatible with the existing version.
+
+## Rejected Alternatives
+
+The main alternative would be switching to an intermediate toolkit to implement both OpenGL and Metal.  The problem with that approach is toolkit size.  Quite a few users cannot abide with the toolkit growing considerably.
+
```

Comment:
I think we are really missing "Considered Alternatives" in terms of completely different approaches. I'm not saying that anything here is not a good way to go, but we should evaluate the other approaches (because they exist).
A few ideas:

* First focus on extracting reusable (style parsing, data loading and handling, data processing) parts 
* Start writing a completely new rendering engine (with the goals above in mind)
* Then integrate the new rendererer with the already extracted parts
---
* Use an abstraction like WebGPU (Metal, Vulkan first and reach OpenGL though translation layers
---
* reuse existing rendering engines and integrate into those

Reply:
Perfectly valid alternatives, to be sure.  But I think they deserve their own proposals.

Reply:
I tried to link and document the discussions around metal here: https://github.com/maplibre/maplibre-gl-native/discussions/317

Reply:
@sjg-wdw it would be great if you could summarize the alternative ideas that were suggested there in say 3 to 5 paragraphs or so.

Reply:
Okay

Reply:
I think the alternatives could be mentioned in this design document, as there is already a section for it.

I do not think that anyone will find the time/motivation to create separate documents outlining those.